c/src/session.rs

```diff
@@ -62,7 +62,59 @@ pub extern "C" fn session_force_close(session: *mut Session) {
     unwrap_void(borrow_mut(session).force_close())
 }
 
+mod private {
+    use std::{ffi::c_void, mem::ManuallyDrop};
+
+    struct SendPtr(*mut c_void);
+    unsafe impl Send for SendPtr {}
+
+    pub(super) struct ForeignCallback<Callback, Finished: FnOnce(*mut c_void)> {
+        data: SendPtr,
+        callback: Callback,
+        finished: ManuallyDrop<Finished>,
+    }
```

Comment:
`*mut c_void` is not `Send`, and neither is a closure that captures one. `SendPtr` provides a way to "smuggle" the pointer between threads. It is up to the user to ensure that it is legal to dereference the pointer in another thread.